File: src/main/java/com/pedigree/editor/commands/MoveNodeCommand.java
package com.pedigree.editor.commands;

import com.pedigree.editor.Command;
import com.pedigree.layout.LayoutResult;

public class MoveNodeCommand implements Command {
    private final LayoutResult layout;
    private final String nodeId;
    private final double oldX;
    private final double oldY;
    private final double newX;
    private final double newY;
    private final Runnable onChanged;

    public MoveNodeCommand(LayoutResult layout, String nodeId, double oldX, double oldY, double newX, double newY, Runnable onChanged) {
        this.layout = layout;
        this.nodeId = nodeId;
        this.oldX = oldX;
        this.oldY = oldY;
        this.newX = newX;
        this.newY = newY;
        this.onChanged = onChanged;
    }

    @Override
    public void execute() {
        layout.setPosition(nodeId, newX, newY);
        if (onChanged != null) onChanged.run();
    }

    @Override
    public void undo() {
        layout.setPosition(nodeId, oldX, oldY);
        if (onChanged != null) onChanged.run();
    }
}

File: src/main/kotlin/com/family/tree/editor/commands/MoveNodeCommand.kt
package com.family.tree.editor.commands

import com.family.tree.editor.Command
import com.family.tree.layout.LayoutResult

class MoveNodeCommand(
    private val layout: LayoutResult,
    private val nodeId: String,
    private val oldX: Double,
    private val oldY: Double,
    private val newX: Double,
    private val newY: Double,
    private val onChanged: Runnable?
) : Command {

    override fun execute() {
        layout.setPosition(nodeId, newX, newY)
        onChanged?.run()
    }

    override fun undo() {
        layout.setPosition(nodeId, oldX, oldY)
        onChanged?.run()
    }
}
